server/resource_server/emails/verification/zk/circuit/mimc_circuit.go

```diff
@@ -0,0 +1,63 @@
+package circuit
+
+import (
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/std/hash/mimc"
+	"github.com/consensys/gnark/std/math/bits"
+	"globe-and-citizen/layer8/server/resource_server/utils"
+)
+
+type MimcCircuit struct {
+	SaltAsVariables  [utils.EmailFrRepresentationSize]frontend.Variable `gnark:",public"`
+	EmailAsVariables [utils.EmailFrRepresentationSize]frontend.Variable `gnark:",secret"`
+
+	VerificationCode [utils.VerificationCodeSize]frontend.Variable `gnark:",public"`
+}
+
+func NewMimcCircuit() *MimcCircuit {
+	return new(MimcCircuit)
+}
+
+func (c *MimcCircuit) Define(api frontend.API) error {
+	mimcInstance, _ := mimc.NewMiMC(api)
+
+	for i := 0; i < utils.EmailFrRepresentationSize; i++ {
+		emailFrVariableBits := bits.ToBinary(api, c.EmailAsVariables[i])
+		saltFrVariableBits := bits.ToBinary(api, c.SaltAsVariables[i])
+
+		// representation for frElementsEmail[i] ^ frElementsSalt[i]
+		currentVariable := frontend.Variable(0)
+		powerOfTwo := frontend.Variable(1)
+		for j := 0; j < len(emailFrVariableBits); j++ {
+			xoredBit := api.Xor(emailFrVariableBits[j], saltFrVariableBits[j])
+			currentVariable = api.Add(currentVariable, api.Mul(xoredBit, powerOfTwo))
+			powerOfTwo = api.Mul(powerOfTwo, 2)
+		}
+
+		mimcInstance.Write(currentVariable)
+	}
+
+	mimcHash := mimcInstance.Sum()
+
+	mimcBits := bits.ToBinary(api, mimcHash)
+
+	code := make([]frontend.Variable, utils.VerificationCodeSize)
+
+	ind := 0
+	for i := 23; i >= 0; i -= 4 {
```

Comment:
This was a challenge I won't lie. You write code at the very edge of my skills in computer science. Learning every day!